core/src/main/java/io/crate/types/DataTypes.java

```diff
@@ -292,4 +292,13 @@ public static void register(int id, DataTypeFactory dataTypeFactory) {
         }
         return streamer;
     }
+
+    public static DataType getFirstNotNullType(List<DataType> dataTypes) {
+        for (DataType dataType : dataTypes) {
+            if (dataType != UNDEFINED) {
+                return dataType;
+            }
+        }
+        return dataTypes.get(0);
```

Comment:
It would fail with an empty list too. Maybe just use Iterables.getFirst  with Iterables.filter